### ISSUE TITLE
Ubuntu network error 5

### ISSUE THREAD

**Issue by amywww** [qtwebkit/qtwebkit#918]
I use qwebkit for SAML authentication. CenOS 7 works well. But Ubuntu 1804 always got some errors. The error is network error 5 for those two URL;

https://aadcdn.msauth.net/ests/2.1/content/cdnbundles/ux.converged.login.strings-en.min_xvnavb8ts_r3tr0w_ckg1g2.js
https://aadcdn.msauth.net/ests/2.1/content/cdnbundles/converged.v2.login.min_mbqre5pw01euigudkiymsa2.css

Use those two urls on Firefox is OK. 

Ubuntu 1604 works fine. 

Can u help?

**Comment by annulen** [comment]
Are you using qtwebkit from Ubuntu repositories?
Could you provide code of minimal project or URL of HTML page which reproduces the error?

**Comment by amywww** [comment]
Hi Annulen,
Thanks for the reply. Here is some detailed info: 
Code for load this authentication url:
m_samlView->page()->mainFrame()->load(m_samlUrl)

m_samlUrl = https://login.microsoftonline.com/66b66353-3b76-4e41-9dc3-fee328bd400e/saml2?SAMLRequest=nZIxT8MwEIX%2FSuTdiR27aWu1lUI7UKmIigQGFuQ419YisYPtIMSvp2lBlKUD4%2Bmevnf37mZetk0n8j4czAO89eBD9NE2xotTY456Z4SVXnthZAteBCWK%2FG4j0piIztlglW1QlHsPLmhrltb4vgVXgHvXCh4fNnN0CKHzIkmMkfoVdzKWn70DPBjgzrogm1jZVnDOkgGdkqTYJvmyQNHqOI42cgD%2FYhq71yZutXLW212wptEGBkKSZVWWsRHDrBpnmAOneForhncALJ1UNScEksE2RdF6NUcvU6rIBHaKjSmv63rMuJoAo6SSnI3Ujh1l3vewNj5IE%2BYoJXSKKcWElykTlAhGn1G0%2FY7hRptam%2F31zKqzyIvbstzi7X1RougJnD%2BteBSgxWyYUJyM3cUtrmPlzwHQ4j9xz5IL08W5%2BvsViy8%3D&RelayState=GwAAAAV%2FwF04ZTRlODQyMDk4OThlNzRlYzhjNjBhM2ZhNzFkMDExNA%3D%3D

After loading this url, on CentOs, we will see user login page and user type user name and password to authenticate. 

But on Ubuntu, only Microsolf logo is displayed, username and password is not shown at all. 

I catch this signal: 
connect (m_pNetworkAccessManager, SIGNAL(finished(QNetworkReply*)),
             this,                    SLOT(onFinished(QNetworkReply*)));
And printout the network error 5 for the two urls I posted in the first page. 

Thanks.

**Comment by annulen** [comment]
Please try setting User-Agent string from old QtWebKit: `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/538.1 (KHTML, like Gecko) QtTestBrowser/0.1 Version/7.0 Safari/538.1`

It fixes error here.

If you are curious, error happens in JS code (exact error message can be seen in inspector or if you print console messages to application output). Seems like site uses `Promise` but implementation in QtWebKit 5.212 is behind current spec, while older QtWebKit doesn't implement promises at all and site falls back to polyfill.

For the record error doesn't reproduce in qtwebkit-dev-wip branch

**Comment by amywww** [comment]
I tried below UserAgent which I printed out from CentOS but not working:
 Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/538.1 (KHTML, like Gecko) XXXUI Version/7.0 Safari/538.1

I will try again and let you know.

**Comment by annulen** [comment]
Indeed, just changing user-agent is insufficient, it seems to require reloading page when inspector is open (and the latter helps even with original UA string). Sorry for misguiding you.

**Comment by amywww** [comment]
Can you be more specific? When should I reload page? After I got the finished signal for those two specific failed urls?

**Comment by annulen** [comment]
>When should I reload page?
I'm afraid this approach won't be possible to automate. I think the only reasonable ways to fix it are
* switch to qtwebkit-dev-wip, or
* try patching away something in JS global evironament to force page into falling back to polyfill (e.g. Promise constructor, but maybe something else)

**Comment by annulen** [comment]
>And what is the error msg you saw on console. 

```
CONSOLE "https://aadcdn.msftauth.net/ests/2.1/content/cdnbundles/ux.converged.login.pcore.min_ancu0urypznffsrky8gjqq2.js:%2" "TypeError: Unable to process binding "component: function (){return { name:'pagination-control',publicMethods:paginationControlMethods,params:{
                            enableCssAnimation:svr.fEnableCssAnimation,disableAnimationIfAnimationEndUnsupported:svr.fDisableAnimationIfAnimationEndUnsupported,initialViewId:initialViewId,currentViewId:currentViewId,initialSharedData:initialSharedData,initialError:$loginPage.getServerError()},event:{
                            cancel:paginationControl_onCancel,showView:$loginPage.view_onShow,setLightBoxFadeIn:view_onSetLightBoxFadeIn,animationStateChange:paginationControl_onAnimationStateChange}} }"
Message: Unable to process binding "foreach: function (){return views }"
Message: Unable to process binding "if: function (){return $parent.currentViewIndex() === $index() }"
Message: Unable to process binding "template: function (){return { nodes:[$data],data:$parent} }"
Message: Unable to process binding "pageViewComponent: function (){return { name:'login-paginated-username-view',params:{
                            serverData:svr,serverError:initialError,isInitialView:isInitialState,displayName:sharedData.displayName,prefillNames:$loginPage.prefillNames,showCredViewBrandingDesc:sharedData.showCredViewBrandingDesc,flowToken:sharedData.flowToken},event:{
                            refresh:$loginPage.view_onRefresh,redirect:$loginPage.view_onRedirect,setPendingRequest:$loginPage.view_onSetPendingRequest,showLearnMore:$loginPage.learnMore_onShow,registerDialog:$loginPage.view_onRegisterDialog,unregisterDialog:$loginPage.view_onUnregisterDialog,showDialog:$loginPage.view_onShowDialog}} }"
Message: a.$index is not a function. (In 'a.$index()', 'a.$index' is an instance of Object)"
```

**Comment by amywww** [comment]
Thanks and I will try both of your suggestions. BTW, when will the new qwebkit release for this JS support?

**Comment by annulen** [comment]
>And where to get .deb for qtwebkit-dev-wip?

Build from sources :)

**Comment by annulen** [comment]
@Emdek could you create PPA for qtwebkit? It could also be used as a dependency for Otter PPA

**Comment by amywww** [comment]
@annulen I am trying the fallback polyfill route but still did not figure it out. I guess we can wait for your   next release. Is there a timeline? Thanks!

**Comment by amywww** [comment]
@annulen Is there any plan to merge this fix to main? Thanks!

**Comment by annulen** [comment]
Which fix? I don't have any.

**Comment by amywww** [comment]
@annulen Are you going to add JS support for promise? This is the fix I was talking about. Thanks.

**Comment by amywww** [comment]
@annulen Can you take a look at my previous comment? Thanks!

**Comment by annulen** [comment]
QtWebKit supports JS promises already

**Comment by amywww** [comment]
@annulen Here is my current Ubuntu qtwebkit info which has the problem:
libqt5webkit5:
  Installed: 5.212.0~alpha2-7ubuntu1
Which version of Ubuntu qtwebkit has the JS promise support you mentioned? Is it the dev version?  Thanks!

**Comment by amywww** [comment]
@annulen Can you let us know if 5.212.0~alpha3 or alpha4 has the support for JS promise? Thanks.

**Comment by amywww** [comment]
@annulen Can you please check if 5.212.0~alpha3 or alpha4 has the support for JS promise? Thanks.

**Comment by Grumbles-hub** [comment]
+1 for the issue here.  In my case it affects being able to use SAML authentication on the Palo Alto GlobalProtect client.  I can verify the version of libqt5webkit5 that ships on Ubutntu 16.04.3 works fine but the version that ships with Linux Mint 20 (libqt5webkit5 5.212.0~alpha4-1ub) still does not work.
